drivers/input/sbutton.c

```diff
@@ -0,0 +1,269 @@
+{
+  FAR struct sbutton_dev_s *priv = (FAR struct sbutton_dev_s *)arg;
+  int                        ret;
+
+  /* Let the event worker know that it has an interrupt event to handle
+   * It is possible that we will already have work scheduled from a
+   * previous interrupt event.  That is OK we will service all the events
+   * in the same work job.
+   */
+
+  if (work_available(&priv->work))
+    {
+      ret = work_queue(HPWORK, &priv->work, sbutton_worker, priv, 0);
+      if (ret != 0)
+        {
+          ierr("ERROR: Failed to queue work: %d\n", ret);
+        }
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sbutton_register
+ *
+ * Description:
+ *   Configure the Single Button Dual Key Keyboard to use the provided
+ *   instance.  This will register the driver as /dev/kbdN where N is the
+ *   minor device number, as well as a joystick at joydevname
+ *
+ * Input Parameters:
+ *   config      - Persistent board configuration data
+ *   kbdminor    - The keyboard input device minor number
+ *
+ * Returned Value:
+ *   Zero is returned on success.  Otherwise, a negated errno value is
+ *   returned to indicate the nature of the failure.
+ *
+ ****************************************************************************/
+
+int sbutton_register(FAR const struct sbutton_config_s *config,
+                     char kbdminor)
+{
+  FAR struct sbutton_dev_s *priv;
+  char                      kbddevname[DEV_NAMELEN];
+  int                       ret;
+
+  /* Debug Sanity Checks */
+
+  DEBUGASSERT(config != NULL);
+  DEBUGASSERT(config->attach != NULL && config->enable != NULL &&
+              config->clear  != NULL);
+
+  priv = kmm_zalloc(sizeof(struct sbutton_dev_s));
+  if (!priv)
+    {
+      ierr("ERROR: kmm_zalloc(%d) failed\n", sizeof(struct sbutton_dev_s));
+      return -ENOMEM;
+    }
+
+  /* Initialize the device driver instance */
+
+  priv->config    = config;  /* Save the board configuration */
+  priv->pressed   = false;
+
+  nxmutex_init(&priv->lock);   /* Initialize device mutex */
+
+  config->clear(config);
+  config->enable(config, false);
+
+  /* Attach the interrupt handler */
+
+  ret = config->attach(config, sbutton_interrupt, priv);
+  if (ret < 0)
+    {
+      ierr("ERROR: Failed to attach interrupt\n");
+      goto errout_with_priv;
+    }
+
+  /* Start servicing events */
+
+  priv->config->enable(priv->config, true);
+
+  snprintf(kbddevname, sizeof(kbddevname), DEV_FORMAT, kbdminor);
+
+  /* Register the device as a keyboard device */
+
+  ret = keyboard_register(&priv->lower, kbddevname,
+                          CONFIG_INPUT_SBUTTON_BUFSIZE);
+  if (ret < 0)
+    {
+      ierr("ERROR: keyboard_register() failed: %d\n", ret);
+      return ret;
```

Comment:
You don't clean up resources on failure here.

Reply:
Good catch!